Adding psi4 and dkh

loriab opened the pull request:
Psi4 is a quantum chemistry package in Python and C++ with pybind11 bindings. It's been conda-packaged for a long time on its own channel (https://anaconda.org/psi4/psi4; recipe https://github.com/psi4/psi4meta/tree/master/conda-recipes/psi4-multiout), so this is finally a port to conda-forge, now that all required dependencies are available. The whole optional dependency ecosystem isn't ported yet, but I'm testing one of every language (C/Py: pylibefp; Fortran: dkh (alongside)), so I think this recipe is stable and ready for review.

The primary concern I anticipate is that this isn't packaging an upstream *release*, only an upstream *branch*. This is because the psi4 conda package and developer build instructions have always been based on `-c defaults -c psi4`, but there's one big and accumulating little changes to prepare for c-f deps, and I'd like to preserve the usability of the existing instructions until the whole software stack is ready on c-f. So if this recipe is merged, I expect that the first release from a c-f feedstock will have an ugly version (e.g., 1.8a2). I'd be glad to have it on a dev branch and `psi4_dev` channel if that can be done at feedstock-creation-time. A c-f compatible psi4 v1.8 release is due for late April/early May.

<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- ~Source is from official source.~ (See notes above)
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- ~If static libraries are linked in, the license of the static library is packaged.~ (No static libs)
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

jturney commented:
I'm willing to be a maintainer for this.

fevangelista commented:
I am happy to help as maintain of this recipe.

isuruf commented on recipes/psi4/conda_build_config.yaml:
```diff
@@ -0,0 +1,23 @@
+channel_sources:
+  - conda-forge/label/libint_dev,conda-forge
+# note for review: I'd engage the following since the src isn't a proper upstream release,
+#   but I don't think it's enough to get this recipe into branch `dev` upon feedstock creation
```
It should be enough

loriab replied:
Thanks! Lines applied, and CI passing again.

carterbox commented on recipes/psi4/conda_build_config.yaml:
```diff
@@ -0,0 +1,21 @@
+channel_sources:
+  - conda-forge/label/libint_dev,conda-forge
+channel_targets:
+  - conda-forge psi4_dev
```
> The primary concern I anticipate is that this isn't packaging an upstream release, only an upstream branch. This is because the psi4 conda package and developer build instructions have always been based on -c defaults -c psi4, but there's one big and accumulating little changes to prepare for c-f deps, and I'd like to preserve the usability of the existing instructions until the whole software stack is ready on c-f. So if this recipe is merged, I expect that the first release from a c-f feedstock will have an ugly version (e.g., 1.8a2). I'd be glad to have it on a dev branch and psi4_dev channel if that can be done at feedstock-creation-time. A c-f compatible psi4 v1.8 release is due for late April/early May.

Using a dev channel is the correct way to address this pre-release problem. Good job.

loriab replied:
Great, thanks. I appreciate these dev channels where one can polish the code and recipes.

carterbox commented on recipes/psi4/conda_build_config.yaml:
```diff
@@ -0,0 +1,21 @@
+  - conda-forge/label/libint_dev,conda-forge
+channel_targets:
+  - conda-forge psi4_dev
+mkl:
+  - 2022.1.0  # [win]
+dftd4:
+  # provides b97-3c parameters
+  - 3.5.0
+libint:
+  # provides rt-switchable ordering (src) and derivative integrals (build)
+  - 2.7.3dev1
+pylibefp:
+  # 0.6.3 can detect py311
+  - 0.6.2
+geometric:
+  # new API
+  - "1.0"
+pymdi:
+  # breaking API change
+  - "1.2"
```
Please only specify dependency versions in the conda_build_config if you want to build against multiple versions. Otherwise, if this package has requires specific versions for dependencies, name them in the meta.yaml.

In fact, we prefer that you don't specify your own constraints if the dependencies are already listed in the [channel-wide pinnings](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml).

loriab replied:
Ok, I moved all the version deps besides `mkl` to meta.yaml. I agree that the windows `mkl` pin is a little troubling -- it was the least disruptive route ("windows blas story").

carterbox commented on recipes/psi4/meta.yaml:
```diff
@@ -0,0 +1,212 @@
+    - if not exist %SP_DIR%\\psi4\\driver\\driver.py exit 1              # [win]
+    # Verify executable
+    - test -f $PREFIX/bin/psi4                                           # [unix]
+    - $PREFIX/bin/psi4 --version                                         # [unix]
+    - $PREFIX/bin/psi4 --module                                          # [unix]
```
```suggestion
    - psi4 --version                                         # [unix]
    - psi4 --module                                          # [unix]
```
This should work without specifying the full path, else the installation is broken.

loriab replied:
I removed the PREFIX from the windows lines, too, and they work fine.

carterbox commented on recipes/psi4/meta.yaml:
```diff
@@ -0,0 +1,212 @@
+    - ldd -v $SP_DIR/psi4/core*.so                                       # [linux]
+    - otool -L $SP_DIR/psi4/core*.so                                     # [osx]
+    - objdump.exe -p %SP_DIR%\\psi4\\core.*.pyd | findstr /i "dll"       # [win]
+    # Actually test
+    - python -c "import psi4;assert psi4.__version__ == '{{ version }}', f'{psi4.__version__=} differs from recipe=\'{{ version }}\''"
+    - python -c "from psi4 import core; core.set_num_threads(42); assert core.get_num_threads() == 42"
+    - MKL_CBWR=AVX pytest -rws -v --durations=20 --color yes -n auto $SP_DIR/psi4 -m "smoke"  # [unix]
+    - pytest -rws -v --durations=20 --color yes -n auto %SP_DIR%\\psi4 -m "smoke and not d2ints"  # [win]
```
```suggestion
    # Actually test
    - pip check
    - python -c "import psi4;assert psi4.__version__ == '{{ version }}', f'{psi4.__version__=} differs from recipe=\'{{ version }}\''"
    - python -c "from psi4 import core; core.set_num_threads(42); assert core.get_num_threads() == 42"
    - MKL_CBWR=AVX pytest -rws -v --durations=20 --color yes -n auto $SP_DIR/psi4 -m "smoke"  # [unix]
    - pytest -rws -v --durations=20 --color yes -n auto %SP_DIR%\\psi4 -m "smoke and not d2ints"  # [win]
```

carterbox commented on recipes/psi4/meta.yaml:
```diff
@@ -0,0 +1,212 @@
+    - scipy
+    # qc
+    - libint {{ libint }}
+    - {{ pin_compatible('optking', upper_bound='0.5') }}
+    - {{ pin_compatible('qcelemental', upper_bound='0.50') }}
+    - {{ pin_compatible('qcengine', upper_bound='0.50') }}
```
There are static linking / overdepending warnings in the logs for these libraries. Can you explain why they are needed at both runtime and build time if they aren't dynamically linked? 

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=690074&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=13902

loriab replied:
These three are pure-python packages with no `run_exports`. Their important role is in `requirements/run`, and the upper bound is because there's some API-breaking changes planned in the next year, and I wanted to isolate the built packages from those. Because these three are required dependencies, if they're not present at cmake configure-time, psi4 will download and vendor them, so their presence in `requirements/host` avoids that.

carterbox replied:
I think you're going to run into unresolvable dependencies in the future if you do it this way because eventually the host environment will move to `optking =0.6.0` and then the run environment pinning expression will become `optking >=0.6.0,<=0.5`?

I think you want something like this:
```yaml
host:
  - optking <=0.5
run:
  - {{ pin_compatible('optking', max_pin='x.x') }}
```
Such that opking is always <=0.5 and the same minor version in the host and run environments.

loriab replied:
Thanks for the analysis. Agreed that the original setup will take a little management of host v run to avoid the situation you describe. optking is fairly new (Dec 2022) and has a much faster release cycle (whenever) than psi4 (twice yearly), so I wanted to leave room for optking bugfixes that psi4 could access (so not pinning to x.x) and also put an upper bound so if optking finds it needs to do a breaking change, this version of psi can be insulated. It's some management, but I'm pretty sure an incompatibility will show up at build-time for me, not at solve-time for user, so I'm ok with it if you are.

loriab replied:
I moved the upper bound for optking out to v0.15, so all three have 12+ versions of space before hitting the limit.

carterbox replied:
I mean this comment to apply to all three of optking, qcelemental, and qcengine. Manually specifying upper_bound is just a very unusual use of these pinning expressions.

> I wanted to leave room for optking bugfixes that psi4 could access (so not pinning to x.x)

Pinning to x.x does allow for bugfixes? For example:

```
host:
  - optking 0.5.2
run:
  - {{ pin_compatible('optking', max_pin='x.x') }}
  # This template would be replaced with optking >=0.5.2,<0.6.0
```
Are the bugfixes for optking not shipped via increment of the patch (major.minor.patch) version number?

carterbox replied:
To allow for the widest compatability (accept bugfixes) whilst protecting against API breaks, you should use max_pin='x.x' for a package whose major version is zero. You can always rebuilt psi4 when new versions of these packages become available.

loriab replied:
I should have said "room for bugfixes and development". optking is new-ish and likely to be fast moving, so I was just being cautious and leaving room for an API breaking change less momentous than 1.0. I can release the constraint if you like.

The qcelemental and qcengine are in a stack with qcfractal (presently at v0.15.8 on c-f). But the last has an imminent API breaking change at 0.50 (currently in development off-c-f at https://anaconda.org/qcarchive/qcfractal/files), and once that's ready, qcelemental and qcengine will have a similar scheduled API break that we'll align at v0.50 for consistency. So I think those two are important to keep.

fwiw, I've used the upper_bound pattern before, and it's worked cleanly for the users (all their complaints come to me :-)  .

carterbox commented on recipes/dkh/meta.yaml:
```diff
@@ -0,0 +1,71 @@
+{% set name = "dkh" %}
```
DKH looks good in both recipe and logs; good job!

loriab commented:
Thanks very much for the review and improvements, @carterbox. I think I've addressed the comments, and I went ahead and added a third-party-licences file here, too, for a few miscellaneous files. CI will be running for a couple hours, but a one-python-version build was clean.

loriab commented:
Sorry to ping you @carterbox, but is there anything I can do to facilitate this PR? I wouldn't usually push, but I've got a release date for psi4 looming. Thanks for all your work reviewing these.

carterbox commented on recipes/psi4/meta.yaml:
```diff
@@ -73,56 +73,57 @@ requirements:
+    - libint 2.7.3dev1                                 # 2.7.3dev1 provides rt-switchable ordering (src) and derivative integrals (build)
+    - {{ pin_compatible('optking', upper_bound='0.15') }}
     - {{ pin_compatible('qcelemental', upper_bound='0.50') }}
     - {{ pin_compatible('qcengine', upper_bound='0.50') }}
```
```suggestion
    - {{ pin_compatible('optking', max_pin='x.x') }}
    - {{ pin_compatible('qcelemental', max_pin='x.x') }}
    - {{ pin_compatible('qcengine', max_pin='x.x') }}
```

carterbox reviewed (approved):
I'm adjusting your pinning templates to be more conservative, so I can merge this PR today. These pinnings will better protect against API breaks. Just rebuild the package whenever new versions of any these dependencies are released.

carterbox commented on recipes/psi4/meta.yaml:
```diff
@@ -0,0 +1,215 @@
+    - scipy
+    # qc
+    - libint 2.7.3dev1                                 # 2.7.3dev1 provides rt-switchable ordering (src) and derivative integrals (build)
+    - {{ pin_compatible('optking', upper_bound='0.15') }}
+    - {{ pin_compatible('qcelemental', upper_bound='0.50') }}
+    - {{ pin_compatible('qcengine', upper_bound='0.50') }}
```
```suggestion
    - {{ pin_compatible('optking', max_pin='x.x') }}
    - {{ pin_compatible('qcelemental', max_pin='x.x') }}
    - {{ pin_compatible('qcengine', max_pin='x.x') }}
```

loriab commented:
> I'm adjusting your pinning templates to be more conservative, so I can merge this PR today. These pinnings will better protect against API breaks. Just rebuild the package whenever new versions of any these dependencies are released.

Sounds good, thanks for the analysis, caution, and approval. :-)
